From c6e111429e38ff123f96758bb0a01229ffe89933 Mon Sep 17 00:00:00 2001
From: Jason Hung <jhung@globalscaletechnologies.com>
Date: Fri, 2 Mar 2018 15:30:59 +0800
Subject: [PATCH 4/6] parser: add preset_ddr_conf field for using preset ddr
 setting

sometimes the ddr configuration which generated by ddr_tool is not
meet our ddr specification or not optimal value, and we don't
want to add these values into ddrparser.pl by hardcode value, so
I add a preset_ddr_conf field on topology file, we can use
pre-define ddr configuration instead of ddr_tool.
---
 ddr/tim_ddr/ddrparser.pl | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/ddr/tim_ddr/ddrparser.pl b/ddr/tim_ddr/ddrparser.pl
index 7206c66..d3850e9 100755
--- a/ddr/tim_ddr/ddrparser.pl
+++ b/ddr/tim_ddr/ddrparser.pl
@@ -6,6 +6,7 @@ sub parse_dram_data
 	my $ddr_type;
 	my $file;
 	my $line;
+	my $preset;
 	my $path = dirname(abs_path($file_name));
 
 	unless (open ($file, "<$file_name")) {
@@ -23,6 +24,10 @@ sub parse_dram_data
 		if ($name =~ m/ddr_type/) {
 			$ddr_type = $value;
 		}
+
+		if ($name =~ m/preset_ddr_conf/) {
+			$preset = $value;
+		}
 	}
 	close($file);
 
@@ -31,10 +36,14 @@ sub parse_dram_data
 	# to change some specific register, we could add "-r" option like below:
 	# ./ddr_tool -i DDR_TOPOLOGY_0.txt -o ddr_static.txt -r 0xC0000380=0x0007A120
 	# ddr_type	DDR3=0, DDR4=1
-	if ($ddr_type eq "0") {
-		system("${path}/ddr3_tool -i $file_name -o ${path}/ddr_static.txt");
-	} elsif ($ddr_type eq "1") {
-		system("${path}/ddr4_tool -i $file_name -o ${path}/ddr_static.txt");
+	if ($preset eq "") {
+		if ($ddr_type eq "0") {
+			system("${path}/ddr3_tool -i $file_name -o ${path}/ddr_static.txt");
+		} elsif ($ddr_type eq "1") {
+			system("${path}/ddr4_tool -i $file_name -o ${path}/ddr_static.txt");
+		}
+	} else {
+		system("/bin/cp ${path}/${preset} ${path}/ddr_static.txt");
 	}
 
 	unlink("${path}/clocks_ddr.txt");
-- 
1.9.1

